Add supporting of SQL functions to AR#select

alextrueman opened the pull request:
### Summary

<!-- Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together. -->

Add support of SQL functions to ActiveRecord::QueryMethods#select with
hashes:
```ruby
Post.joins(:comments).select('COUNT(comments.id)' => :comment_count)
```

### Other Information

Regarding to discussion on this PR: https://github.com/rails/rails/pull/45612

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

<!--
Note: Please avoid making *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.
Create a pull request when it is ready for review and feedback
from the Rails team :).
-->

fschwahn commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1864,7 +1864,11 @@ def transform_select_hash_values(fields)
             arel_column(key.to_s) do
-              predicate_builder.resolve_arel_attribute(klass.table_name, key.to_s)
+              if (/(\w+)\s*\(/).match?(key.to_s)
+                Arel.sql(key.to_s)
```
I think the point of the discussion in #45612 was that this does not happen automatically, but the user has to knowingly opt-in by calling `Arel.sql` themselves. 

So the user has to do 
```ruby
Post.joins(:comments).select(Arel.sql('COUNT(comments.id)') => :comment_count).take
```
to make it abundantly clear that they pass in raw sql.

alextrueman replied:
@fschwahn, then I see no point in this PR as user can simply write it in this way:
```ruby
Post.joins(:comments).select(Arel.sql("COUNT(comments.id)").as("comment_count")).take
```
with a current solution 🤔

byroot replied:
Yeah, I confirm that I don't feel confident allowing unmarked arbitrary SQL in there.

Also the implementation using a regexp to try to guess whether it's a function is not robust enough at all. Not all expressions are function calls, e.g. `Model.select('something + 1')` is a totally valid expression.

@simi if you still want the `Arel.sql` key thing, feel free to submit a PR and tag me.

alextrueman committed:
Add supporting of SQL functions to AR#select

Add support of SQL functions to ActiveRecord::QueryMethods#select with
hashes:

  Post.joins(:comments).select('COUNT(comments.id)' => :comment_count)

Co-authored-by: Josef Šimánek <193936+simi@users.noreply.github.com>
Co-authored-by: Jean byroot Boussier <19192189+casperisfine@users.noreply.github.com>

byroot commented:
Closing this as I don't think this is the way.
